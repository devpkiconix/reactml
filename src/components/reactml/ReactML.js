import React from 'react';
import {
    __, defaultTo, path as pathGet,
    curry,
} from 'ramda';
import mapValues from 'lodash/mapValues';
import dependencies from './dependencies';
import renderLib from './render';
import defaultTagFactory from './materialUiTagFactory';

const renderer = curry(renderLib.render)(dependencies);

const { connect, withStyles } = dependencies;

const defaultToEmpty = defaultTo({})

const state2PropsMaker = (compName, props) => {
    const state2propsPath = ['spec', 'components', compName, 'state-to-props'];
    const stateNodeName = props.spec.state.stateNodeName;

    const oState2pathSpec = defaultToEmpty(
        pathGet(state2propsPath, props)
    );
    return (state) => {
        const imPathGet =
            (dotted) => state.reactml.getIn([stateNodeName].concat(dotted.split('.')));
        const mapped = mapValues(oState2pathSpec, imPathGet);
        // console.log(state.reactml.toJS(), "mapped", mapped);
        return mapped;
    };
};

export const ReactML = (props) => {
    const stateNodeName = props.stateNodeName;
    const compName = props.component;
    const stylesPath = ['spec', 'components', compName, 'styles'];
    const viewNodePath = ['spec', 'components', compName, 'view']
    const mapStateToProps = state2PropsMaker(compName, props);
    const actionExtractor = () => props.actionLib;
    const compProps = {
        tagFactory: props.tagFactory || defaultTagFactory,
        root: pathGet(viewNodePath, props),
        // ...mapStateToProps,
        stateNodeName,
    };
    return React.createElement(
        connect(mapStateToProps, actionExtractor())(
            withStyles(
                defaultToEmpty(
                    pathGet(stylesPath, props)
                ))(
                    renderer
                )
        ),
        compProps);
};

