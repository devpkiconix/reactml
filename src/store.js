import { createStore, applyMiddleware, compose } from 'redux'
import { connectRouter, routerMiddleware } from 'connected-react-router'
import thunk from 'redux-thunk'
import logger from 'redux-logger'
import createHistory from 'history/createBrowserHistory'
import rootReducer from './modules'


export const history = createHistory()

const initialState = {}
const enhancers = []
const middleware = [thunk, routerMiddleware(history)]

if (process.env.NODE_ENV === 'development') {
  const devToolsExtension = window.__REDUX_DEVTOOLS_EXTENSION__

  middleware.push(logger)
  if (typeof devToolsExtension === 'function') {
    enhancers.push(devToolsExtension())
  }
}

const composedEnhancers = compose(
  applyMiddleware(...middleware),
  ...enhancers
)


function makeStore() {
  const store = createStore(
    connectRouter(history)(rootReducer),
    initialState,
    composedEnhancers
  );
  return store;
}

export default makeStore;