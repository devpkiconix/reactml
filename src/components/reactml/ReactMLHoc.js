import React from 'react';
import { connect } from 'react-redux';
import { ReactML } from '../../components/reactml/ReactML';

const ReactmlHoc = (spec, actionLib, tagFactory, stateNodeName, view) => {
    class View extends React.Component {
        componentDidMount() {
            // If initial state has been provided by YAML
            // and not already been initialized, then
            // generate a redux action to init state
            if (spec.state.initial &&
                this.props.stateInitialized === false) {

                this.props.dispatch({
                    type: 'REACTML_INIT', initial: {
                        ...spec.state.initial,
                        spec, actionLib, tagFactory,
                        stateNodeName,
                    },
                    stateNodeName,
                });
            }
        }

        render() {
            return (<ReactML
                tagFactory={tagFactory}
                stateNodeName={stateNodeName}
                spec={spec} component={view}
                actionLib={actionLib}
            />);
        }
    }

    const mapStateToProps = (state) => {
        const reactml = state.reactml;
        let stateInitialized = false;
        // If the state has already been initialized,
        // set a flag so we don't have to init on every HOC
        if (reactml.get(stateNodeName)) {
            stateInitialized = true;
        }
        return {stateInitialized};
    };
    return connect(mapStateToProps)(View);
}


export default ReactmlHoc;
