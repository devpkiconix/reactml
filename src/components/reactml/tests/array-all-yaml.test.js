import React from 'react';
import { Provider } from 'react-redux'
import { ReactML } from '../ReactML';
import ReactMLArrayMapper from '../ArrayMapper';
import ReactMLHoc from '../ReactMLHoc';
import renderer from 'react-test-renderer';
import store from './store';

const tagFactory = {};
const lib = {};

const initial = {
    todoList: [
        { label: "get milk", done: true },
        { label: "buy beans", done: true },
        { label: "grind beans", done: true },
        { label: "brew coffee", done: false },
        { label: "boil milk", done: false },
        { label: "mix'em up", done: false },
        { label: "enjoy", done: false },
    ]
};

describe("Uninterpreted props", () => {
    beforeEach(() => {
        return store.dispatch({
            type: 'REACTML_INIT',
            initial,
            stateNodeName: 'todoApp'
        });
    });

    it("array rendering with uninterpreted props", () => {
        const spec = {
            state: {
                stateNodeName: 'todoApp',

                // NOTE: State initialization is not tested here. That would require
                // multiple redux actions to be processed asynchronously, and seems
                // overly complex to test. For now, we'll test the case where the
                // state has already been initialized.
            },
            components: {
                Todo: {
                    view: {
                        tag: 'div',
                        children: [
                            {
                                tag: 'input',
                                props: {
                                    type: 'checkbox',
                                    value: '.todo.done',
                                },
                            },
                            {
                                tag: 'span',
                                content: '.todo.label',
                            }
                        ]
                    },
                },
                todoListView: {
                    'state-to-props': {
                        todoList: 'todoList'
                    },
                    view: {
                        tag: 'div',
                        children: [
                            {
                                tag: 'h4', content: 'Todo List',
                            },
                            {
                                tag: 'ReactMLArrayMapper',
                                props: {
                                    over: '.todoList',
                                    as: 'todo',
                                    component: '...Todo'
                                }
                            }
                        ]
                    }
                }

            }
        };
        const actionLib = {};
        const tagFactory = { ReactMLArrayMapper, };
        const compName = 'todoListView';
        const stateNodeName = 'todoApp';

        const TodoApp = ReactMLHoc(spec,
            actionLib, tagFactory,
            stateNodeName, compName);

        const elem = renderer
            .create(
                <Provider store={store}>
                    <TodoApp />
                </Provider>);

        const tree = elem.toJSON();
        expect(tree).toMatchSnapshot();
    });
})
