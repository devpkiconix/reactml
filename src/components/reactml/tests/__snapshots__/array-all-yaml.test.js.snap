// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Uninterpreted props array rendering with uninterpreted props 1`] = `
<div>
  <h4>
    Todo List
  </h4>
  <div>
    <div>
      <input
        type="checkbox"
        value={true}
      />
      <span>
        get milk
      </span>
    </div>
    <div>
      <input
        type="checkbox"
        value={true}
      />
      <span>
        buy beans
      </span>
    </div>
    <div>
      <input
        type="checkbox"
        value={true}
      />
      <span>
        grind beans
      </span>
    </div>
    <div>
      <input
        type="checkbox"
        value={false}
      />
      <span>
        brew coffee
      </span>
    </div>
    <div>
      <input
        type="checkbox"
        value={false}
      />
      <span>
        boil milk
      </span>
    </div>
    <div>
      <input
        type="checkbox"
        value={false}
      />
      <span>
        mix'em up
      </span>
    </div>
    <div>
      <input
        type="checkbox"
        value={false}
      />
      <span>
        enjoy
      </span>
    </div>
  </div>
</div>
`;
